interfaces/system_key.go

```diff
@@ -378,11 +384,80 @@ func RemoveSystemKey() error {
 }
 
 func MockSystemKey(s string) func() {
-	var sk systemKey
-	err := json.Unmarshal([]byte(s), &sk)
+	sk, err := SystemKeyFromString(s)
 	if err != nil {
 		panic(err)
 	}
-	mockedSystemKey = &sk
+	mockedSystemKey = sk.(*systemKey)
 	return func() { mockedSystemKey = nil }
 }
+
+type SystemKeyMismatchAction int
+
+const (
+	SystemKeyMismatchActionUndefined SystemKeyMismatchAction = iota
+	SystemKeyMismatchActionNone
+	SystemKeyMismatchActionRegenerateProfiles
+)
+
+func (s SystemKeyMismatchAction) String() string {
+	switch s {
+	case SystemKeyMismatchActionNone:
+		return "none"
+	case SystemKeyMismatchActionRegenerateProfiles:
+		return "regenerate-profiles"
+	default:
+		return fmt.Sprintf("<undefined %d>", int(s))
+	}
+}
+
+var (
+	ErrSystemKeyMismatchVersionTooHigh = errors.New("system-key in version higher than supported")
+)
+
+// SystemKeyMismatchAdvise checks the provided and currently saved system keys
+// to device whether security profiles should be regenerated. Returns
+// ErrSystemKeyMismatchVersionTooHigh when the provided system key is newer than
+// one supported by the current process.
+func SystemKeyMismatchAdvise(maybeOther any) (SystemKeyMismatchAction, error) {
+	other, ok := maybeOther.(*systemKey)
+	if !ok {
+		return SystemKeyMismatchActionUndefined, fmt.Errorf("internal error: not a system key")
+	}
+
+	// system-key is regeneraterd on startup of snapd, so anything read back
+	// from disk should match what currently exeuting snapd supports
+	my, err := readSystemKey()
+	if err != nil {
+		return SystemKeyMismatchActionUndefined, err
+	}
+
+	if other.Version == my.Version {
+		// same version as our key, let's double check the mismatch, as the
+		// client may have generated a system key right right before snapd
+		// startup, so they did not observe the latest content of the key
+		match, err := SystemKeysMatch(my, other)
+		if err != nil {
+			// unreachable
+			return SystemKeyMismatchActionUndefined, err
+		}
+
+		if match {
+			return SystemKeyMismatchActionNone, nil
+		}
+	} else if other.Version < systemKeyVersion {
+		// fallback behavior for lower versions of system key observed by the
+		// client, most likely the client is older than the current snapd
+		// process, selectively compare keys that have special meaning
+		if other.NFSHome == my.NFSHome {
+			// client's view of NFS home is different then ours, let's reload
```

Comment:
this comment sounds wrong vs the code below

Reply:
updated